larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+{
+    return m_pStrongClassifier->Predict(features);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateProbability(const DoubleVector &features) const
+{
+    return (m_pStrongClassifier->Predict(features) + 1.0) * 0.5;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void AdaBoostDecisionTree::ReadXmlFile(const std::string &bdtFileName, const std::string &bdtName)
+{
+    m_isInitialized = true;
+
+    pandora::TiXmlDocument xmlDocument(bdtFileName);
+
+    if (!xmlDocument.LoadFile())
+    {
+        std::cout << "AdaBoostDecisionTree::Initialize - Invalid xml file." << std::endl;
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_FAILURE);
+    }
+
+    const pandora::TiXmlHandle xmlDocumentHandle(&xmlDocument);
+    pandora::TiXmlNode *pContainerXmlNode(pandora::TiXmlHandle(xmlDocumentHandle).FirstChildElement().Element());
+
+    // Find the xml container for the bdt with the required name
+    while (pContainerXmlNode)
+    {
+        if (pContainerXmlNode->ValueStr() != "AdaBoostDecisionTree")
+            throw pandora::StatusCodeException(pandora::STATUS_CODE_FAILURE);
+
+        const pandora::TiXmlHandle currentHandle(pContainerXmlNode);
+
+        std::string currentName;
+        PANDORA_THROW_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, pandora::XmlHelper::ReadValue(currentHandle, "Name", currentName));
+
+        if (currentName.empty() || (currentName.size() > 1000))
+        {
+            std::cout << "AdaBoostDecisionTree::Initialize - Implausible bdt name extracted from xml." << std::endl;
+            throw pandora::StatusCodeException(pandora::STATUS_CODE_INVALID_PARAMETER);
+        }
+
+        if (currentName == bdtName)
+            break;
+
+        pContainerXmlNode = pContainerXmlNode->NextSibling();
+    }
+
+    if (!pContainerXmlNode)
+    {
+        std::cout << "AdaBoostDecisionTree: Could not find an AdaBoostDecisionTree of name " << bdtName << std::endl;
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_NOT_FOUND);
+    }
+
+    // Read the components of this bdt container
+    pandora::TiXmlHandle localHandle(pContainerXmlNode);
+    pandora::TiXmlElement *pCurrentXmlElement = localHandle.FirstChild().Element();
+
+    while (pCurrentXmlElement)
+    {
+        if (pandora::STATUS_CODE_SUCCESS != this->ReadComponent(pCurrentXmlElement))
+        {
+            std::cout << "AdaBoostDecisionTree: Unknown component in xml file" << std::endl;
+            throw pandora::StatusCodeException(pandora::STATUS_CODE_FAILURE);
+        }
+
+        pCurrentXmlElement = pCurrentXmlElement->NextSiblingElement();
+    }
+    m_pStrongClassifier = new StrongClassifier(m_weakClassifiers);
```

Comment:
See previous comments. When is `m_pStrongClassifier` deallocated?

Can we avoid the `m_weakClassifiers` member variable and just keep containers of weak classifiers on the local stack, passed via reference between functions, until this allocation is done, or does this not fit later usage?

Reply:
Done.